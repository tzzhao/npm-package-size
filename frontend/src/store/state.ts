import {PackageInformation} from 'npm-pkg-utils';

export enum PackageState {
  INIT, LOADING, READY, ERROR
}

export interface RootState {
  state: PackageState,
  packageInfos: PackageInformation[],
  globalError: Error | undefined
}

export const initialState: RootState = {
  state: PackageState.INIT,
  packageInfos: [],
  globalError: undefined
};
