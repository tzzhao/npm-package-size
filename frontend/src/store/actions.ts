import {PackageInformation} from 'npm-pkg-utils';
import {PackageState} from './state';

export const SET_LOADING = 'SET_LOADING';
export const SET_PACKAGE_INFOS = 'SET_PACKAGE_INFOS';
export const SET_ERROR = 'SET_ERROR';

export const LoadingAction = () => ({
  type: SET_LOADING,
  payload: {
    state: PackageState.LOADING,
    globalError: undefined,
    packageInfos: []
  }
});

export const SetPackageInformationAction = (packageInfos: PackageInformation[]) => ({
  type: SET_PACKAGE_INFOS,
  payload: {
    state: PackageState.READY,
    globalError: undefined,
    packageInfos
  }
});

export const SetErrorAction = (globalError: Error) => ({
  type: SET_ERROR,
  payload: {
    state: PackageState.ERROR,
    globalError,
    packageInfos: []
  }
});
