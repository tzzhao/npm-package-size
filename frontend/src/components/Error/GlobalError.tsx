import React from 'react';
import {RootState} from "../../store/state";
import {connect} from "react-redux";

interface ErrorProperties {
  error: Error | undefined;
}

let GlobalError: React.FC<Partial<ErrorProperties>> = props => {
  if (!props.error) return null;
  return (
      <div className="global-error">
        The following error occurred: [{props.error?.name}] {props.error?.message ? props.error.message: ''}
      </div>
  )
};

const mapStateToProps = (state: RootState) => {
  return {
    error: state.globalError
  };
};

GlobalError = connect(mapStateToProps)(GlobalError);

export default GlobalError;
