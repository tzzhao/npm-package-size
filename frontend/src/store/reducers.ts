import {Reducer} from 'redux';
import {Action, SET_ERROR, SET_LOADING, SET_PACKAGE_INFOS} from './actions';
import {initialState, RootState} from './state';

export const reducer : Reducer<RootState, Action> = ((state: RootState = initialState, action: Action) => {
  switch(action.type) {
    case SET_LOADING:
      return {...state, ...action.payload};
    case SET_PACKAGE_INFOS:
      return {...state, ...action.payload};
    case SET_ERROR:
      return {...state, ...action.payload};
    default:
      return state;
  }
});
