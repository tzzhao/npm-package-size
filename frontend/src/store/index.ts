import {createStore, Store} from 'redux';
import {reducer} from './reducers';
import {initialState} from './state';

export const store: Store = createStore(reducer, initialState);
