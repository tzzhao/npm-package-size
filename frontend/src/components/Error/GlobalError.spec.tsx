import {render, RenderResult} from '@testing-library/react';
import * as React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import {createStore, Store} from 'redux';
import {reducer} from '../../store/reducers';
import {initialState} from '../../store/state';
import {
  TEST_ERROR_MESSAGE,
  TEST_ERROR_NAME,
  TestLoadingAction,
  TestSetErrorAction,
  TestSetPackageInformationAction,
} from '../../utils/test/store/actions';
import {GlobalError} from './GlobalError';

const store: Store = createStore(reducer, initialState);

test('Test error is empty on init', () => {
  const renderResult: RenderResult = render(<Provider store={store}><GlobalError /></Provider>);
  expect(renderResult.baseElement.textContent).toBe("");
});

test('Test error is empty during loading', () => {
  store.dispatch(TestLoadingAction);
  const renderResult: RenderResult = render(<Provider store={store}><GlobalError /></Provider>);
  expect(renderResult.baseElement.textContent).toBe("");
});

test('Test error is empty on ready', () => {
  store.dispatch(TestSetPackageInformationAction);
  const renderResult: RenderResult = render(<Provider store={store}><GlobalError /></Provider>);
  expect(renderResult.baseElement.textContent).toBe("");
});

test('Test error is displayed properly on global error', () => {
  store.dispatch(TestSetErrorAction);
  const renderResult: RenderResult = render(<Provider store={store}><GlobalError /></Provider>);
  expect(renderResult.baseElement.textContent).toBe(`The following error occurred: [${TEST_ERROR_NAME}] ${TEST_ERROR_MESSAGE}`);
});

test('Test error is displayed properly on global error - SNAPSHOT', () => {
  store.dispatch(TestSetErrorAction);
  const component = renderer.create(<Provider store={store}><GlobalError /></Provider>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
