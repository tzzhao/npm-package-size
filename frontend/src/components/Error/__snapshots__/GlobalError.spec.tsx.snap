// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test error is displayed properly on global error - SNAPSHOT 1`] = `
<div
  className="global-error"
>
  The following error occurred: [
  GlobalError
  ] 
  [global error message]
</div>
`;
